torch/nn/utils/stateless.py

```diff
@@ -37,21 +36,52 @@ def _setattr(self, name: str, value: Any) -> None:
     module._orig_class = cls
 
 
-def _check_tied_val_already_replaced(old_val, new_val, replaced_tensors_map):
-    if old_val not in replaced_tensors_map:
-        replaced_tensors_map[old_val] = new_val
-    elif replaced_tensors_map[old_val] is not new_val:
-        warnings.warn("functional_call was passed multiple values for tied weights. "
-                      "This behavior is deprecated and will be an error in future versions")
+def _create_tied_weights_map(module, params_and_buffers):
+    # creates a weight map of {tied_name: name_given_by_user} for all weights where one of their tied weights is passed
+    #
+    # The basic algorithm looks like:
+    #   - index all weights by their original tensor value to find tied weights
+    #     - when we encounter a weight not used by the user, we save it in a set (second element in the tuple)
+    #     - when we run into a weight used by the user, we save that separate from the set as the first element in the tuple
+    #     - ending map looks like {tensor: (name_given_by_user, set(all_tied_names)}
+    #   - then loop through the values of this map (name_given_by_user and set(all_tied_names))
+    #     - for each element of all_tied_names, add {tied_name: name_given_by_user} to a new map
+
+    names = params_and_buffers.keys()
+    weight_to_name_and_tied_names: Dict[torch.Tensor, Tuple[Optional[str], Set[str]]] = {}
+
+    def add_to_name_map(name, t):
+        if t in weight_to_name_and_tied_names:
+            first_seen_name = weight_to_name_and_tied_names[t][0]
+            if name in names and first_seen_name and params_and_buffers[name] is not params_and_buffers[first_seen_name]:
+                raise ValueError(f"functional_call got values for both {name} and {first_seen_name}, which are tied.")
+            elif name in names:
+                weight_to_name_and_tied_names[t] = (name, weight_to_name_and_tied_names[t][1])
+            else:
+                weight_to_name_and_tied_names[t][1].add(name)
+        else:
+            weight_to_name_and_tied_names[t] = (name, set()) if name in names else (None, {name})
```

Comment:
nit: This (the nested if-else, and the repeat of the name in names condition) is a bit difficult to read. Maybe a suggestion?

```py
def add_to_name_map(name, t):
    if t not in weight_to_name_and_tied_names:
        weight_to_name_and_tied_names[t] = (name, set()) if name in names else (None, {name})
        return
    if name not in names:
        weight_to_name_and_tied_names[t][1].add(name)
        return
    first_seen_name = weight_to_name_and_tied_names[t][0]
    if first_seen_name is None:
        weight_to_name_and_tied_names[t] = (name, weight_to_name_and_tied_names[t][1])
        return
    if params_and_buffers[name] is not params_and_buffers[first_seen_name]:
        raise ValueError("...")
    weight_to_name_and_tied_names[t][1].add(name)
```
Not sure if this is better but it took me a while to unwrap what was going on

Reply:
Updated with more comments too. But it's confusing, I agree. Hopefully this refactor and the doc comments help